Common/Extensions.cs

```diff
@@ -1445,5 +1445,39 @@ public static string RemoveFromEnd(this string s, string ending)
                 return s;
             }
         }
+
+        /// <summary>
+        /// Queues the specified work to run on the thread pool that must complete within a specified time interval.
+        /// If the work completes within the interval, a new action is invoked if defined.
+        /// </summary>
+        /// <param name="action">The work to execute asynchronously</param>
+        /// <param name="delay">The time span to wait before completing the returned task, or TimeSpan.FromMilliseconds(-1) to wait indefinitely.</param>
+        /// <param name="continuationAction">The work to execute synchronously after the first work is completed</param>
+        /// <exception cref="TimeoutException">The exception that is thrown when the time allotted for the asynchronously work has expired.</exception>
+        public static async Task TimeoutAfterAndContinue(this Action action, TimeSpan? delay, Action continuationAction = null)
+        {
+            using (var timeoutCancellationTokenSource = new CancellationTokenSource())
+            {
+                var task = Task.Run(action, timeoutCancellationTokenSource.Token);
+                var delayTask = Task.Delay(delay ?? TimeSpan.FromMilliseconds(-1), timeoutCancellationTokenSource.Token);
+
+                if (delayTask == await Task.WhenAny(task, delayTask))
+                {
+                    timeoutCancellationTokenSource.Cancel();
+                    if (delay.HasValue)
+                    {
+                        throw new TimeoutException(
+                            $"The operation has timed out after {delay.Value.TotalMinutes:F2} minutes"
+                        );
+                    }
+                }
+
+                timeoutCancellationTokenSource.Cancel();
+
+                await task;
```

Comment:
Wondering if at this stage wouldn't `task` already be completed?

Reply:
Correct, since we did `await Task.WhenAll(task, delayTask)`

Reply:
If the task is completed within timeout, it may have faulted or been canceled. So we need to re-await the task so that any exceptions/cancellation is rethrown. Ref.: https://stackoverflow.com/a/11191070
I will add a comment to document the re-wait.